### ISSUE TITLE
Service Application Credentials

### ISSUE THREAD

**Issue by samuelfelt** [athenahealth/apiserver-athenaFlex#177]
The API documentations indicates that there are three different forms of authentication:
```
Authentication
The API Server supports end user authentication through integration with Microsoft Azure Active Directory (AAD) services. Three classes of users are supported:
1. Providers and Staff
2. Patients and Authorized Representatives
3. Application Services
```

1 and 2 seem to be well documented, but I'm curious about 3. Based on our use case that seems like the most appropriate. 
The most I can find is the below, but I can't find any documentation on how the provider organization should provision that access.
```
Service Applications are authorized by Client Credentials Grant  defined in OAuth 2.0. 
The authorization process for the application is handled offline by the provider organization. 
Basically, the provider organization will create a login with the API Server User role enabling the application to be authorized to the system. 
The provider organization must then share credentials created for that application and the application must be configured to use those credentials.
```

We host a separate application that host its own login and authorization, we are looking to sync appointment and patient data into that application for sending appointment reminders. As such we will not be authorizing based on either a patient or provider level. Organizations as a whole opt-in or out.

**Comment by Torin-Shepard-athena** [comment]
Please see Application Service Authentication on the Overview > Security page at https://mydata.athenahealth.com/fhirapidoc.
Some applications log in to the API server directly to support their operations. The API provides servers such as these the capability to use a single login to manage the information they need to access through CPS or CEMR with minimal application changes. This method may be used by applications providing portal services, secure messaging, patient facing registration or scheduling capabilities, etcetera.

All API calls require a user authenticated by AAD with a corresponding user configured in the local AD used for direct login to CPS/CEMR/athenaPractice/athenaFlow.  This requirement exists because EHR user permissions are used to enforce proper RBAC.

Owners of Service Applications must request EHR administrator to create a user account in both AAD and the local system.  EHR administrator will grant appropriate permissions so this account is authorized to access all the data the Service Application will read/write using the APIs.  The EHR administrator will share the credentials for this account so they can be configured in the Service Application.  Although this account could technically be used to log into the EHR client application, it does not correspond to any actual person. Both the EHR administrator and Service App owner must safeguard the credentials to prevent unintended/undesired use of this account to gain access to data stored in the EHR.

**Comment by AjeetYadav07** [comment]
Closing this issue, if you have any query please raise a new ticket.

**Comment by Torin-Shepard-athena** [comment]
It is possible to get access tokens using "native user" authorization flow. The attached document explains how to use this flow in V20 and earlier product versions using GET from token endpoint with grant_type=client_credentials and service application user name and password in HTTP Authorization header.

[Notes on athenaPractice Native User Auth Flow.docx](https://github.com/athenahealth/apiserver-athenaFlex/files/6902309/Notes.on.athenaPractice.Native.User.Auth.Flow.docx)

In future product versions, this authorization flow will be changed to require POST to token endpoint with grant_type=password, service app user credentials in HTTP request body, and client id and secret in HTTP Authorization header.  More information will be available with the next major product release.

**Comment by stonstad** [comment]
Hello.  Is the native user auth workflow still supported?  I'm looking to follow the above directions and I am confirming this is an option for us.  Thanks!

**Comment by stonstad** [comment]
> In future product versions, this authorization flow will be changed to require POST to token endpoint with grant_type=password, service app user credentials in HTTP request body, and client id and secret in HTTP Authorization header. More information will be available with the next major product release.

Has this workflow been modified to require POST?

**Comment by stonstad** [comment]
Confirming the above workflow works using GET.  Use ap22sandbox.fhirapi.athenahealth.com.

**Comment by AjeetYadav07** [comment]
Yes, Native user auth workflow using GET and POST both are supported in ap22sandbox but in future product release auth workflow using GET will be removed.

**Comment by AjeetYadav07** [comment]
@stonstad,
Marking this issue as closed. If you have any query, please raise a new ticket.
